### PR TITLE
Add support for PyPy.

### PR DESCRIPTION
**Pull request by jamadden**
This needs ZODB 4.2.0b1. 

Aside from some small performance tweaks to ClientStorage, the changes here are mostly to deal with PyPy's non-refcounted GC and consist of things like closing files in a timely manner.

### CODE REVIEW

**Comment by jamadden** [comment]
All builds except the Python 2.6 build passed. It failed with the following error; notably weird is the fact that it refers to Python 2.7 in the path. I restarted the build in case it was transient.

```python
Error in test check_connection_management (ZEO.tests.testConnection.MonitorTests)
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 318, in run
    self.setUp()
  File "/home/travis/build/zopefoundation/ZEO/src/ZEO/tests/ConnectionTests.py", line 111, in setUp
    self.startServer()
  File "/home/travis/build/zopefoundation/ZEO/src/ZEO/tests/ConnectionTests.py", line 221, in startServer
    sconf, zconf, addr[1], keep)
  File "/home/travis/build/zopefoundation/ZEO/src/ZEO/tests/forker.py", line 177, in start_zeo_server
    s.connect(adminaddr)
  File "/usr/lib/python2.7/socket.py", line 224, in meth
    return getattr(self._sock,name)(*args)
error: [Errno 111] Connection refused
```

**Comment by jamadden** [comment]
This time Python 2.6 failed with two errors. Is there any known flakiness in Python 2.6? Could this possibly be the result of https://github.com/zopefoundation/ZEO/commit/fa2b97d47f5763e32140e7321a0500f8b3100fc6?

```python
Error in test checkTimeoutProvokingConflicts (ZEO.tests.testConnection.FileStorageTimeoutTests)
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 327, in run
    testMethod()
  File "/home/travis/build/zopefoundation/ZEO/src/ZEO/tests/ConnectionTests.py", line 1130, in checkTimeoutProvokingConflicts
    storage.tpc_finish(t)
  File "/home/travis/build/zopefoundation/ZEO/src/ZEO/ClientStorage.py", line 1196, in tpc_finish
    tid = self._server.tpc_finish(id(txn))
  File "/home/travis/build/zopefoundation/ZEO/src/ZEO/ServerStub.py", line 252, in tpc_finish
    return self.rpc.call('tpc_finish', id)
  File "/home/travis/build/zopefoundation/ZEO/src/ZEO/zrpc/connection.py", line 776, in call
    r_args = self.wait(msgid)
  File "/home/travis/build/zopefoundation/ZEO/src/ZEO/zrpc/connection.py", line 796, in wait
    raise DisconnectedError()
DisconnectedError
```

```python
Failure in test checkQuickVerificationWith2Clients (ZEO.tests.testConnection.FileStorageInvqTests)
Traceback (most recent call last):
  File "/usr/lib/python2.7/unittest/case.py", line 327, in run
    testMethod()
  File "/home/travis/build/zopefoundation/ZEO/src/ZEO/tests/ConnectionTests.py", line 625, in checkQuickVerificationWith2Clients
    self.assertEqual(perstorage.verify_result, "quick verification")
  File "/usr/lib/python2.7/unittest/case.py", line 511, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/usr/lib/python2.7/unittest/case.py", line 504, in _baseAssertEqual
    raise self.failureException(msg)
AssertionError: 'empty cache' != 'quick verification'
```

**Comment by jamadden** [comment]
Something is wrong with the use of `virtualenv` in .travis.yml. We're getting the wrong python versions. *Every* build is running under Python 2.7.3. Here's output from the purported 2.6 build:

```
$ python --version
Python 2.6.9
$ pip --version
pip 6.0.7 from /home/travis/virtualenv/python2.6.9/lib/python2.6/site-packages (python 2.6)
...
$ env/bin/python --version
Python 2.7.3
```

And the purported 3.4 build:
```
$ python --version
Python 3.4.2
$ pip --version
pip 6.0.7 from /home/travis/virtualenv/python3.4.2/lib/python3.4/site-packages (python 3.4)
$ virtualenv env
New python executable in env/bin/python
...
$ env/bin/python bootstrap.py
...
$ env/bin/python --version
Python 2.7.3
```

**Comment by tseaver** [comment]
ISTM that the `.travis.yml` is bogus:  it uses `virtualenv` w/o qualification. It should be running something like `python -m virtualenv` instead (or just use `tox` instead of `zc.buildout`).

**Comment by jamadden** [comment]
I think the use of `zc.buildout` is to be able to use `zope.testrunner` and run tests in parallel with `bin/test -v1 -j99`. Luckily, it seems to work just fine without an explicit virtualenv.

**Comment by jamadden** [comment]
Sigh. Python 2.7 passes, Python 2.6 failed with another DisconnectedError, and all versions of Python 3 (now that we're actually testing Python 3) fail with the following:

```
Exception raised:
    Traceback (most recent call last):
      File "/opt/python/3.2.5/lib/python3.2/doctest.py", line 1288, in __run
        compileflags, 1), test.globs)
      File "<doctest ZEO.tests.testZEO2.some_basic_locking_tests[29]>", line 1, in <module>
        pprint.pprint(zs1.server_status(), width=40)
      File "/home/travis/build/zopefoundation/ZEO/src/ZEO/StorageServer.py", line 757, in server_status
        return self.server.server_status(self.storage_id)
      File "/home/travis/build/zopefoundation/ZEO/src/ZEO/StorageServer.py", line 1313, in server_status
        self.storages[storage_id].lastTransaction().encode('hex'))
    AttributeError: 'bytes' object has no attribute 'encode'
```

Also, all the Python3 tests got killed because they ran for 15 minutes and produced no output in the last ten, so something appears to be hanging.

PyPy failed because some of the tests have a dependency (I think) on the iteration order of dictionaries. I guess I forgot about those.

**Comment by jamadden** [comment]
Python 2.6, 2.7 and PyPy are green. Python 3.x at least runs now, and I fixed the new nagios/ruok protocol. However, it's tests are broken due to `u` prefixes in the doctests. I'm not sure how that's usually fixed?

**Comment by mgedmin** [comment]
> However, its tests are broken due to u prefixes in the doctests. I'm not sure how that's usually fixed?

Often with a [custom doctest output checker](https://github.com/zopefoundation/zope.browsermenu/blob/53147af82158031e6f7cd040c98adad7d1ba0a45/src/zope/browsermenu/tests/test_menu.py#L22).

**Comment by jamadden** [comment]
Thanks, I didn't realize you could do match group substitutions.

**Comment by jamadden** [comment]
Yay, everything is green except for Python 3.2. It failed with the following error. I think there's a race condition because I've seen each of the builds fail with that error at different times. I'll kick 3.2 off again and see if it goes this time.

```python
Failure in test checkQuickVerificationWith2Clients (ZEO.tests.testConnection.FileStorageInvqTests)
Traceback (most recent call last):
  File "/opt/python/3.2.5/lib/python3.2/unittest/case.py", line 370, in _executeTestPart
    function()
  File "/home/travis/build/zopefoundation/ZEO/src/ZEO/tests/ConnectionTests.py", line 625, in checkQuickVerificationWith2Clients
    self.assertEqual(perstorage.verify_result, "quick verification")
  File "/opt/python/3.2.5/lib/python3.2/unittest/case.py", line 628, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/python/3.2.5/lib/python3.2/unittest/case.py", line 1049, in assertMultiLineEqual
    self.fail(self._formatMessage(msg, standardMsg))
  File "/opt/python/3.2.5/lib/python3.2/unittest/case.py", line 494, in fail
    raise self.failureException(msg)
AssertionError: 'empty cache' != 'quick verification'
- empty cache
+ quick verification
```

**Comment by jamadden** [comment]
This time Python 3.4 failed in `checkQuickVerificationWith2Clients`. I'm not seeing what the non-deterministic path is, but it's definitely there. 

A re-run "fixed" the issue and the build is green across the board.

**Comment by jamadden** [comment]
Ping.

File: src/ZEO/ClientStorage.py [hunk 1]
```diff
@@ -19,6 +19,7 @@
 
 """
 import BTrees.IOBTree
+import gc
```
Comment (tseaver) [inline src/ZEO/ClientStorage.py]: Unused?

**Comment by tseaver** [comment]
Released:

- https://pypi.python.org/pypi/ZEO/4.2.0b1

**Pull request merged**